core/src/main/java/co/aikar/commands/BaseCommand.java

```diff
@@ -64,69 +65,228 @@
+     * You will have to manually write to {@link #commandName}.
+     */
+    public BaseCommand() {
+    }
 
-    public BaseCommand() {}
-    public BaseCommand(String cmd) {
+    public BaseCommand(@Nullable String cmd) {
         this.commandName = cmd;
     }
 
+    /**
+     * Takes a string like "foo|bar baz|qux" and generates a list looking as such:
+     * - foo baz
+     * - foo qux
+     * - bar baz
+     * - bar qux
+     * <p>
+     * This is applied to every possible sub command combination.
+     *
+     * @param subCommandParts
+     *         All parts of the subcommand (i.e. foo, bar, baz, and qux).
+     *
+     * @return List of all sub command possibilities.
+     */
+    private static Set<String> getSubCommandPossibilityList(String[] subCommandParts) {
+        int i = 0;
+        Set<String> current = null;
+        while (true) {
+            Set<String> newList = new HashSet<>();
+
+            if (i < subCommandParts.length) {
+                for (String s1 : ACFPatterns.PIPE.split(subCommandParts[i])) {
+                    if (current != null) {
+                        newList.addAll(current.stream().map(s -> s + " " + s1).collect(Collectors.toList()));
+                    } else {
+                        newList.add(s1);
+                    }
+                }
+            }
+
+            if (i + 1 < subCommandParts.length) {
+                current = newList;
+                i = i + 1;
+                continue;
+            }
+
+            return newList;
+        }
+    }
+
     /**
      * Gets the root command name that the user actually typed
+     *
      * @return Name
      */
     public String getExecCommandLabel() {
         return execLabel;
     }
 
+    /**
+     * Gets the actual args in string form the user typed
+     *
+     * @return Args
+     */
+    public String[] getOrigArgs() {
+        return origArgs;
+    }
+
     /**
      * Gets the actual sub command name the user typed
+     *
      * @return Name
      */
     public String getExecSubcommand() {
         return execSubcommand;
     }
 
     /**
-     * Gets the actual args in string form the user typed
-     * @return Args
+     * This returns a list of all tab complete options which are possible with the given argument and commands.
+     *
+     * @param arg
+     *         Argument which was pressed tab on.
+     * @param cmds
+     *         The possibilities to return.
+     *
+     * @return All possible options. This may be empty.
      */
-    public String[] getOrigArgs() {
-        return origArgs;
+    private static List<String> filterTabComplete(String arg, List<String> cmds) {
+        return cmds.stream()
+                   .distinct()
+                   .filter(cmd -> cmd != null && (arg.isEmpty() || ApacheCommonsLangUtil.startsWithIgnoreCase(cmd, arg)))
+                   .collect(Collectors.toList());
     }
 
+    /**
+     * This sets the parenting command of this.
```

Comment:
i think this method is unused. the wording is super confusing. we need to jus remove this setter as itll never be API nor should it change.